{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Resource Quantification Estimation\n",
    "Here is a simple method for resource quantification using numerical simulation. It leverages the fact that the only way for mRNA or protein to \"disappear\" from a system is via degradation/dilution.\n",
    "\n",
    "## Energetic Costs of Individual mRNAs and Proteins\n",
    "The cost of transcription and translation can be estimated by the amount of phosphate bonds involved in the biosynthesis of each nucleotide or amino acid used as well as for the elongation process. These were determined in experiments by Lynch and Marinov in their paper, [Bioenergetic Costs of a Gene](https://doi.org/10.1073/pnas.1514974112).\n",
    "\n",
    "### Transcription\n",
    "An approximate average for the synthesis of nucleoside triphosphates (NTPs) is 46 phosphate bonds, with an additional 2 for polymersiation for a total of 48 phosphate bonds. This is summarised in the following equation, with $\\epsilon$ representing the length of the gene sequence in nucleotides.\n",
    "\n",
    "\\begin{align*}\n",
    "    \\text{cost}_{m_i} = \\epsilon \\cdot 48\n",
    "\\end{align*}\n",
    "\n",
    "### Translation\n",
    "The costs for translation are more complicated with each amino acid having varying biosynthesis costs. The approach is to multiply the number of each amino acid in a protein sequence by its associated cost and sum over all amino acids. \n",
    "\n",
    "\\begin{align*}\n",
    "    \\text{cost}_{x_i} = \\sum_{AA} \\text{number of AA} \\cdot \\text{cost of AA}\n",
    "\\end{align*}\n",
    "\n",
    "The phosphate costs of each amino acid in the table below:\n",
    "| **Amino Acid** | **One-letter symbol** | **~P Cost** |\n",
    "|----------------|-----------------------|-------------|\n",
    "| Ala            | A                     | 11.7        |\n",
    "| Cys            | C                     | 24.7        |\n",
    "| Asp            | D                     | 12.7        |\n",
    "| Glu            | E                     | 15.3        |\n",
    "| Phe            | F                     | 52.0        |\n",
    "| Gly            | G                     | 11.7        |\n",
    "| His            | H                     | 38.3        |\n",
    "| Ile            | I                     | 32.3        |\n",
    "| Lys            | K                     | 30.3        |\n",
    "| Leu            | L                     | 27.3        |\n",
    "| Met            | M                     | 34.3        |\n",
    "| Asn            | N                     | 14.7        |\n",
    "| Pro            | P                     | 20.3        |\n",
    "| Gln            | Q                     | 16.3        |\n",
    "| Arg            | R                     | 27.3        |\n",
    "| Ser            | S                     | 11.7        |\n",
    "| Thr            | T                     | 18.7        |\n",
    "| Val            | V                     | 23.3        |\n",
    "| Trp            | W                     | 74.3        |\n",
    "| Tyr            | Y                     | 50.0        |\n",
    "\n",
    "- The total cost of expressing a gene over the life of an organism can be estimated as the individual cost multiplied by the concentration determined from a numerical simulation. \n",
    "- We discretise the ODEs and change the differential of a species into difference equations. \n",
    "- To calculate the total amount of a species synthesis, we utilise the fact that degradation/dilution terms are linear and that the concentration of a species is calculated at every time point. \n",
    "- In this manner, we can simply calculate the amount of a species that has been degraded/diluted (in terms of concentration) to the concentration of the species at steady-state added to the initial concentration to obtain the total amount synthesised (in terms of concentration). \n",
    "- Finally, we can multiply the concentration by the volume to obtain the amount. In the case of \\emph{E. coli}, the volume can be regarded at $1 \\mu M$. \n",
    "\n",
    "After numerical simulation, we can extract the time-series data of a species, $A_i$, from the solution vector. This is a sequence of numbers, \\{$\\mathbf{A}_i$\\}. \n",
    "\\begin{align*}\n",
    "    \\mathbf{A}_i &= \\begin{bmatrix}\n",
    "        A_i [0] \\\\\n",
    "        A_i [1] \\\\\n",
    "        \\vdots \\\\\n",
    "        A_i [N] \\\\\n",
    "      \\end{bmatrix}\n",
    "\\end{align*}\n",
    "\n",
    "Where $N$ is the number of time points simulated.\n",
    "\n",
    "The calculation the total number of an mRNA ($m_i$) or protein ($x^\\omega$) species simulated up to time, $t$, is as follows:\n",
    "\\begin{align*}\n",
    "    \\dot{m}_i [t] &= m_i [t] - m_i [t-1] = k_{\\text{eff}, 1; i} \\cdot g^T_i - \\delta_i \\cdot m_i [t] \\\\\n",
    "    m_i [t] &= \\sum_1^{\\tau = t} (k_{\\text{eff}, 1; i} \\cdot g^T_i - \\delta_i \\cdot m_i [\\tau]) + m_i[0]\\\\\n",
    "    m_{i,\\text{ synthesised}} [t] &=  k_{\\text{eff}, 1; i} \\cdot g^T_i \\cdot t + m_i[0] \\\\\n",
    "    &= \\sum_1^{\\tau = t} ( k_{\\text{eff}, 1; i} \\cdot g^T_i - \\delta_i \\cdot m_i [\\tau] ) + \\sum_1^{\\tau = t} \\delta_i \\cdot m_i [\\tau] + m_i[0] \\\\\n",
    "    &= m_i [t] + m_i[0] + \\delta_i \\cdot \\sum_1^{\\tau = t} m_i [\\tau]\\\\\n",
    "    x^{\\omega}_{i,\\text{ synthesised}} [t] &= x^{\\omega}_i [t] + x^{\\omega}_i [0] + \\lambda \\cdot \\sum_1^{\\tau = t} x^{\\omega}_i [\\tau]\n",
    "\\end{align*}\n",
    "\n",
    "Species are denoted as discrete functions of time to reflect the numerical simulation. This is signified by the square brackets around the variable, $t$. The amount synthesised over the number of time points simulated, $N$, would be $m_{i, \\text{ synthesised}} [N]$ and $x^{\\omega}_{i,\\text{ synthesised}} [N]$. When additional reactions that involve a species are present, we can simply perform a linear addition of the synthesis terms of the additional related species.\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
